File: lollys-lotto-ts-sdk/src/codegen/instructions/requestWinningNumbersV2.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface RequestWinningNumbersV2Accounts {
  authority: PublicKey
  lottoGame: PublicKey
  randomnessAccountData: PublicKey
  eventEmitter: PublicKey
  /** The Solana System program. Used to allocate space on-chain for the randomness_request account. */
  systemProgram: PublicKey
}

export function requestWinningNumbersV2(
  accounts: RequestWinningNumbersV2Accounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: false, isWritable: false },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: false },
    {
      pubkey: accounts.randomnessAccountData,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([94, 19, 50, 41, 216, 5, 2, 148])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: programs/lollys-lotto/src/instructions/switchboard/request_winning_numbers_v2.rs
use anchor_lang::prelude::*;

use crate::{errors::LollysLottoError, pda_identifier::PDAIdentifier, state::{EventEmitter, LottoGame}};
use switchboard_on_demand::RandomnessAccountData;


#[derive(Accounts)]
pub struct RequestWinningNumbersV2<'info> {
    /// CHECK: Authority of the LottoGame instance
    pub authority: AccountInfo<'info>,

    #[account(
        has_one = authority,
        seeds = [
            LottoGame::IDENT,
            authority.key().as_ref(),
            lotto_game.load()?.round.to_le_bytes().as_ref(),
        ],
        bump = lotto_game.load()?.bump,
    )]
    pub lotto_game: AccountLoader<'info, LottoGame>,

    /// CHECK: The account's data is validated manually within the handler.
    pub randomness_account_data: AccountInfo<'info>,


    #[account(mut)]
    pub event_emitter: Box<Account<'info, EventEmitter>>,

    /// The Solana System program. Used to allocate space on-chain for the randomness_request account.
    pub system_program: Program<'info, System>,

}


pub fn request_winning_numbers_v2(
    ctx: Context<RequestWinningNumbersV2>,
) -> anchor_lang::prelude::Result<()> {
let clock: Clock = Clock::get()?;
    let randomness_data = RandomnessAccountData::parse(ctx.accounts.randomness_account_data.data.borrow()).unwrap();

    if randomness_data.seed_slot != clock.slot - 1 {
        msg!("seed_slot: {}", randomness_data.seed_slot);
        msg!("slot: {}", clock.slot);
        return Err(LollysLottoError::OnDemandRandomnessNotResolved.into());
    }

    Ok(())
}